feat: Add TableTools hooks & components

bastilian opened the pull request:
This moves the hooks and components used in compliance to build tables and toolbars into the utilities package to make them available to other applications.

To test the functionality there is draft PR on the compliance repository that removes everything that will be moved in here and replaces the imports properly: https://github.com/RedHatInsights/compliance-frontend/pull/1519

Hyperkid123 reviewed (commented):
@bastilian as this feature includes react components, could you please move the files to `/packages/component`. I know we talk about utilities but I thought we were only talking about some hooks.

I will do a proper review later on.

bastilian commented:
@Hyperkid123 The components are now moved to the components package. I'm not really sure why Travis is failing though. Locally it builds normally and tests pass.

Hyperkid123 commented:
@bastilian it still references files in the utils package:
```
error TS5055: Cannot write file '/home/travis/build/RedHatInsights/frontend-components/packages/utils/useTableTools/FilterConfigBuilder/ChipBuilder.js' because it would overwrite input file.
```

bastilian commented:
@Hyperkid123 Thanks for the hint! What do I need to do with imports like these: https://github.com/RedHatInsights/frontend-components/pull/1359/files#diff-bbfaa02c4e2b592a8b584507c51a22eec6b2e1c14c82166a28d9f9ca12e45b7eR1 Are this not allowed/possible or is there a proper way to do them?

Hyperkid123 commented:
@bastilian I have run out of time I will check it tomorrow and see what the issue is.

bastilian commented:
@Hyperkid123 yeah, no worries! Thank you so far!

Hyperkid123 commented:
@bastilian looks like you have fixed it, there are just linter issues in the travis job

Hyperkid123 reviewed (changes requested):
This reexport is incorrect. Referenced file does not exists + you should only export files from your mini-module.

You should basically move all your `useHooks` move to separate directories in order to create a mini-module from each. The directory structure should look something like this

```
components
|- useBulkSelect
|    |- index.js
|    |- useBulkSelect.js
|- useColumnManager
|    |- index.js
|    |- useColumnManager.js
...
|- index.ts # all new mini-modules must be exported here as well
```

That is how the packages should be structured from now on. Some old modules are different but that is because we don't want to introduce any breaking changes. But we will eventually migrate them.

bastilian commented:
@Hyperkid123 Thank you so much! I'll update the directory structure and go over the imports again.

karelhala commented on packages/utils/src/useTableTools/FilterConfigBuilder/FilterConfigBuilder.js:
```diff
@@ -0,0 +1,232 @@
+    };
+  };
+
+  removeFilterWithChip = (chips, activeFilters) => {
```
Since all `getItemByLabelOrValue`, `getCategoryForLabel` and `removeFilterFromGroupFilterState` can be moved away from the class let's move this one as well.

karelhala commented on packages/utils/src/useTableTools/FilterConfigBuilder/FilterConfigBuilder.js:
```diff
@@ -0,0 +1,232 @@
+    const chipCategory = chips.category;
+    const chipLabel = chips.chips[0].name;
+    const chipItem = this.getItemByLabelOrValue(chipLabel, chipCategory);
+    const chipValue = chipItem ? chipItem.value : chipLabel;
```
Let's use safe accessor in here
```suggestion
    const chipValue = chipItem?.value : chipLabel;
```

karelhala reviewed (commented):
Thank you @bastilian for the submission! These features are long requested by a few teams and I think it would be awesome to share them between projects.

The amount of review comments is not because I'd hate the code, on contrary! The code looks good, it's just tailored to specific usecase so let's make it sharable and configurable by other apps as well!

I haven't got time to look at the hooks of this PR, but let's go in small steps. First let's reconsider usage of classes and focus on the hooks in second step.

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+    }
+  };
+  const onSave = () => {
+    onSaveArgument && onSaveArgument(selectedColumns);
```
This is incorporated. I only recently realised that optional chaining works like that too. :D

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+
+const ColumnManager = ({ isOpen, columns, onSave: onSaveArgument, onClose: onCloseProp, selectedColumns: selectedColumnsProp = [] }) => {
+  const [selectedColumns, setSelectedColumns] = useState(selectedColumnsProp);
+  const isSelectedColumn = ({ title }) => selectedColumns.includes(title);
```
The `isSelectedColumn` function is moved into a helpers module.

The `title` is really not the best. It's now changed to use the key.

Additionally the key is a required prop for the column prop.
If no key is given it will not show up in the manager, this is also to encourage using a key prop for columns.

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+  };
+  const toggleColumn = (column) => {
+    if (isSelectedColumn(column)) {
+      setSelectedColumns(selectedColumns.filter((title) => title !== column.title));
```
Updated.

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+    if (isSelectedColumn(column)) {
+      setSelectedColumns(selectedColumns.filter((title) => title !== column.title));
+    } else {
+      setSelectedColumns([...selectedColumns, column.title]);
```
Same as above.

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+  };
+  const onClose = () => {
+    setSelectedColumns(selectedColumnsProp);
+    onCloseProp && onCloseProp();
```
Updated.

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+        Select all
+      </Button>
+    </TextContent>
+  );
```
It's now a proper component.

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+        <Button key="cancel" variant="secondary" onClick={onClose}>
+          Cancel
+        </Button>,
+      ]}
```
Both can now be set via the props `saveLabel` and `closeLabel`.

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+    <Modal
+      title="Manage columns"
+      isOpen={isOpen}
+      variant="small"
```
There is now a `modalProps` prop  to pass in props.

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+
+  return (
+    <Modal
+      title="Manage columns"
```
The title is now set via a `title` prop.

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+            isSelected={isSelectedColumn(column)}
+            onChange={() => {
+              toggleColumn(column);
+            }}
```
Changed.

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+            onChange={() => {
+              toggleColumn(column);
+            }}
+            isDisabled={idx === 0}
```
There is now a `isRequired` prop that columns can have, which will disable the column in the manager.

For now this should do it, but we should take a look at the column definitions. At the moment they are just as Patternfly needs them and props mixed in for certain hooks, like sorting. It's a bit of a chaos.

bastilian commented on packages/components/src/TableToolsTable/TableToolsTable.js:
```diff
@@ -0,0 +1,52 @@
+import TableToolbar from '../TableToolbar';
+import useTableTools from '@redhat-cloud-services/frontend-components-utilities/useTableTools';
+
+const TableToolsTable = ({ items = [], columns = [], filters = [], options = {}, toolbarProps: toolbarPropsProp, ...tablePropsRest }) => {
```
Added all of them.

bastilian replied:
I thought about this again, and I think I'll remove them again. I think of this Table component being sortof default table for the platform and there should be no need to extensively customise it. If there is a new feature or change to be made to any table it should be made here rather than as a customisation, to mitigate the chance of inconsistencies across applications. I will also look into ways of not needing  the custom toolbar and table props, these were added as shortcuts/workarounds and are reason for a few not so nice parts in the code to deal with them. 

Should there be a reason for a different component body, for example one that doesn't have a PrimaryToolbar or maybe one migrating to the new PF composable tables, only the `useTableTools` hook can be used to invoke all hooks at once or the underlying hooks can also be used individually with Patternfly (based) Table components, for example the Compliance `SystemsTable` only uses a few hooks, like the `useFilterConfig` hook and `useBulkSelect` hook with the `InventoryTable`

bastilian commented on packages/components/src/TableToolsTable/__fixtures__/columns.js:
```diff
@@ -0,0 +1,52 @@
+"use strict";
```
This was just a bad copy of already compiled code. :D Should be good now.

bastilian commented on packages/utils/src/useTableTools/FilterConfigBuilder/FilterBuilder.js:
```diff
@@ -0,0 +1,47 @@
+    return filterStringArray.map((fs) => fs.join(' or ')).join(moreThanTwo ? ' and ' : '');
+  };
+
+  buildFilterString = (filters) => {
```
This filter builder can probably go back into Compliance as helper functions. I'll look into that. 

The idea is/was that filters have a function/config to prepare the filter values to query an API, but this can be implemented to be more generic, I believe.

bastilian replied:
This is now removed from here and will be implemented in Compliance itself.

bastilian commented on packages/utils/src/useTableTools/helper.js:
```diff
@@ -0,0 +1,63 @@
+export const filterSelected = (items, selectedIds = []) => items.filter((item) => selectedIds.includes(item.itemId));
+
+export const filteredAndSortedItems = (items, filter, sorter) => {
+  const filtered = filter ? filter(items) : items;
```
Updated.

bastilian commented on packages/utils/src/useTableTools/helper.js:
```diff
@@ -0,0 +1,63 @@
+
+export const filteredAndSortedItems = (items, filter, sorter) => {
+  const filtered = filter ? filter(items) : items;
+  return sorter ? sorter(filtered) : filtered;
```
Updated.

bastilian commented on packages/utils/src/useTableTools/helper.js:
```diff
@@ -0,0 +1,63 @@
+};
+
+const mergeIfArray = (firstValue, secondValue) => {
+  if (firstValue?.constructor.toString().indexOf('Array') > -1) {
```
Updated.

bastilian commented on packages/utils/src/useTableTools/helper.js:
```diff
@@ -0,0 +1,63 @@
+};
+
+export const mergeFilters = (currentFilters, additionalFilters) =>
+  Object.keys(currentFilters).reduce((acc, filter) => {
```
Updated.

bastilian commented on packages/utils/src/useTableTools/helper.js:
```diff
@@ -0,0 +1,63 @@
+    return property(item);
+  } else {
+    return item[property];
+  }
```
Updated.

bastilian commented on packages/utils/src/useTableTools/helper.js:
```diff
@@ -0,0 +1,63 @@
+  }
+};
+
+export const stringToId = (string) => string.split(' ').join('').toLowerCase();
```
Updated.

bastilian commented on packages/utils/src/useTableTools/helper.js:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  return fallback;
+};
```
Updated.

bastilian commented on packages/utils/src/useTableTools/helper.js:
```diff
@@ -0,0 +1,63 @@
+    .split(/[-_\W]+/g)
+    .map((string) => string.trim())
+    .map((string) => string[0].toUpperCase() + string.substring(1))
+    .join('');
```
Updated.

bastilian commented on packages/utils/src/useTableTools/useItemIdentify.js:
```diff
@@ -0,0 +1,25 @@
+
+// TODO Rethink how items are identified
+// this hook is meant to give items in the table a reproducible unique ID in order to identify them mainly for the bulk selection
+export const useItemIdentify = (items, options = {}) => {
```
Moved it to the helpers.

bastilian commented on packages/utils/src/useTableTools/useToolbarActions.js:
```diff
@@ -0,0 +1,15 @@
+const prependEmptyFirstAction = (actions) => [undefined, ...actions];
+
+const useToolbarActions = ({ actions: actionsOption = [] }) => {
```
Changed it to a helper function.

bastilian commented on packages/components/src/SelectedFilterSwitch/SelectedFilterSwitch.js:
```diff
@@ -0,0 +1,14 @@
+
+const SelectedFilterSwitch = ({ isChecked, setActiveFilter }) => (
+  <Switch label="Selected only" isChecked={isChecked} onChange={() => setActiveFilter('selected', !isChecked)} />
+);
```
Updated.

bastilian commented on packages/components/src/ColumnManager/ColumnManagerListItem.js:
```diff
@@ -0,0 +1,30 @@
+import propTypes from 'prop-types';
+import { DataListCheck, DataListControl, DataListItem, DataListItemRow, DataListCell, DataListItemCells } from '@patternfly/react-core';
+
+const ColumnManagerListItem = ({ column, onChange, isSelected, isDisabled, id }) => (
```
Updated.

bastilian commented on packages/components/src/ColumnManager/ColumnManagerListItem.js:
```diff
@@ -0,0 +1,30 @@
+import { DataListCheck, DataListControl, DataListItem, DataListItemRow, DataListCell, DataListItemCells } from '@patternfly/react-core';
+
+const ColumnManagerListItem = ({ column, onChange, isSelected, isDisabled, id }) => (
+  <DataListItem>
```
Updated.

bastilian commented on packages/components/src/ColumnManager/ColumnManagerListItem.js:
```diff
@@ -0,0 +1,30 @@
+  <DataListItem>
+    <DataListItemRow>
+      <DataListControl>
+        <DataListCheck checked={isSelected} onChange={onChange} otherControls isDisabled={isDisabled} id={`${id}-id`} name={`${id}-id`} />
```
Updated.

bastilian commented on packages/components/src/ColumnManager/ColumnManagerListItem.js:
```diff
@@ -0,0 +1,30 @@
+      </DataListControl>
+      <DataListItemCells
+        dataListCells={[
+          <DataListCell key={`${id}-cell`}>
```
Updated.

bastilian commented on packages/components/src/ColumnManager/ColumnManagerListItem.js:
```diff
@@ -0,0 +1,30 @@
+          </DataListCell>,
+        ]}
+      />
+    </DataListItemRow>
```
Updated.

bastilian commented on packages/components/src/ColumnManager/ColumnManager.js:
```diff
@@ -0,0 +1,78 @@
+
+ColumnManager.propTypes = {
+  columns: propTypes.array,
+  selectedColumns: propTypes.array,
```
There is now a proper props type definition for a column.

karelhala commented on packages/utils/src/useBulkSelect/helpers.js:
```diff
@@ -0,0 +1,40 @@
+
+export const itemIds = (items) => items.map((item) => item.itemId);
+
+export const selectItemTransformer = (item, selectedIds) => ({
+  ...item,
+  rowProps: {
+    selected: (selectedIds || []).includes(item.itemId),
+  },
+});
```
Could be better to use default value in here, rather then in each computation
```suggestion
export const selectItemTransformer = (item, selectedIds = []) => ({
  ...item,
  rowProps: {
    selected: selectedIds.includes(item.itemId),
  },
});
```

bastilian replied:
Added.

karelhala commented on packages/utils/src/useBulkSelect/useBulkSelectWithItems.js:
```diff
@@ -0,0 +1,48 @@
+  const items = propItems.map((item) => selectItemTransformer(item, preselected));
+  const total = items.length;
+
+  const filteredItems = filter ? filter(items) : items;
```
I'd simplify this by using null comparator
```suggestion
  const filteredItems = filter?.(items) ?? items;
```

bastilian replied:
Added.

karelhala commented on packages/utils/src/useBulkSelect/useBulkSelectWithItems.js:
```diff
@@ -0,0 +1,48 @@
+  const filteredTotal = filteredItems.length;
+  const filtered = filteredTotal !== total;
+
+  const paginatedItems = paginator ? paginator(filteredItems) : filteredItems;
```
Same as above
```suggestion
  const paginatedItems = paginator?.(filteredItems) ?? filteredItems;
```

bastilian replied:
Added.

karelhala commented on packages/utils/src/useBulkSelect/useBulkSelectWithItems.js:
```diff
@@ -0,0 +1,48 @@
+  const paginatedItems = paginator ? paginator(filteredItems) : filteredItems;
+  const paginatedTotal = paginatedItems.length;
+
+  const allCount = filtered ? filteredTotal : total;
```
And here
```suggestion
  const allCount = filteredTotal ?? total;
```

bastilian replied:
Added.

karelhala commented on packages/utils/src/useBulkSelect/useBulkSelectWithItems.js:
```diff
@@ -0,0 +1,48 @@
+
+  const allCount = filtered ? filteredTotal : total;
+
+  const setPageMemo = useMemo(() => setPage, []);
```
I think that this one should be `useCallback` it's basically the same, but `useMemo` is generally used for heavy computations.

karelhala commented on packages/utils/src/useColumnManager/helper.js:
```diff
@@ -0,0 +1,8 @@
+const managableColumns = (columns) =>
+  columns.map((column) => (column?.managable === undefined ? { ...column, managable: true } : column)).filter((column) => column.managable === true);
+
+export const filterManageableColumns = (columns, prop) => managableColumns(columns.filter((column) => column[prop] !== undefined));
```
I would use safe accessor in here
```suggestion
export const filterManageableColumns = (columns = [], prop = '') => managableColumns(columns.filter((column) => column?.[prop] !== undefined));
```

karelhala commented on packages/utils/src/useColumnManager/helper.js:
```diff
@@ -0,0 +1,8 @@
+
+export const filterColumnsBySelected = (columns, selected, selectProp) => columns.filter((column) => selected.includes(get(column, selectProp)));
+
+const managableColumns = (columns) =>
```
I would rather pass in here a default value
```suggestion
const managableColumns = (columns = []) =>
```

karelhala commented on packages/utils/src/useColumnManager/helper.js:
```diff
@@ -0,0 +1,8 @@
+import get from 'lodash/get';
+
+export const filterColumnsBySelected = (columns, selected, selectProp) => columns.filter((column) => selected.includes(get(column, selectProp)));
```
We should use default values for helper functions, you never know who's going to use them
```suggestion
export const filterColumnsBySelected = (columns = [], selected = false, selectProp = '') => columns.filter((column) => selected.includes(get(column, selectProp)));
```

karelhala commented on packages/utils/src/useColumnManager/useColumnManager.js:
```diff
@@ -0,0 +1,43 @@
+    setIsManagerOpen(false);
+  };
+
+  return enableColumnManager && managableColumns.length !== 0
```
You can use safe accessor here
```suggestion
  return managableColumns?.length !== 0
```

karelhala commented on packages/utils/src/useDedicatedAction/useDedicatedAction.js:
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+
+// TODO Refactor.
+const useDedicatedAction = (options) => {
```
This feels more like a helper function rather then hook. Can we omit the `use` from the name of the function?

karelhala commented on packages/utils/src/useDedicatedAction/useDedicatedAction.test.js:
```diff
@@ -0,0 +1,21 @@
+
+describe('useDedicatedAction', () => {
+  it('returns a dedicated action toolbar config', () => {
+    const { result } = renderHook(() =>
```
The `useDedicatedAction` is not a hook, so no need to use `renderHook`

karelhala commented on packages/utils/src/useExport/helpers.js:
```diff
@@ -0,0 +1,55 @@
+
+const encoding = (format) => `data:${ENCODINGS[format]};charset=utf-8`;
+
+export const linkAndDownload = (data, filename) => {
```
There's already [download file](https://github.com/RedHatInsights/frontend-components/blob/master/packages/utils/src/helpers/helpers.js#L13) utils function

karelhala commented on packages/utils/src/useExport/helpers.js:
```diff
@@ -0,0 +1,55 @@
+  return encodeURI(`${encoding('json')},${JSON.stringify(result)}`);
+};
+
+export const callCallback = (callback, ...args) => callback && callback(...args);
```
Trying to understand the reasoning for this helper function. Much easier would be to use directly `callback?.()` I don't think we really need it, do we?

karelhala commented on packages/utils/src/useExport/useExport.js:
```diff
@@ -0,0 +1,40 @@
+const useExport = ({ exporter, columns = [], isDisabled = false, onStart, onComplete, onError }) => {
+  const exportColumns = exportableColumns(columns);
+  const exportWithFormat = async (format) => {
+    callCallback(onStart);
```
Safe accessor can be used instead of `callCallback`
```suggestion
    onStart?.();
```

karelhala commented on packages/utils/src/useExport/useExport.js:
```diff
@@ -0,0 +1,40 @@
+    callCallback(onStart);
+    const items = await exporter()
+      .then((items) => {
+        callCallback(onComplete, items);
```
Safe accessor can be used instead of callCallback
```suggestion
        onComplete?.(items);
```

karelhala commented on packages/utils/src/useExport/useExport.js:
```diff
@@ -0,0 +1,40 @@
+  const exportColumns = exportableColumns(columns);
+  const exportWithFormat = async (format) => {
+    callCallback(onStart);
+    const items = await exporter()
+      .then((items) => {
```
No need to combine `await` and `then` if you want to catch errors you can always wrap things up with `try`/`catch`

karelhala commented on packages/utils/src/useExport/useExport.js:
```diff
@@ -0,0 +1,40 @@
+        callCallback(onComplete, items);
+        return items;
+      })
+      .catch((error) => callCallback(onError, error));
```
Safe accessor can be used instead of `callCallback`
```suggestion
      .catch((error) => onError?.(error));
```

karelhala commented on packages/utils/src/useExport/useExport.js:
```diff
@@ -0,0 +1,40 @@
+import { callCallback, filename, linkAndDownload, csvForItems, jsonForItems, exportableColumns } from './helpers';
+
+const useExport = ({ exporter, columns = [], isDisabled = false, onStart, onComplete, onError }) => {
```
Since this function doesn't use any of the React's lifecycle let's not call it `useExport` it's just a helper function after all.

karelhala commented on packages/utils/src/useExport/useExportWithItems.js:
```diff
@@ -0,0 +1,20 @@
+
+import useExport from './useExport';
+
+export const useExportWithItems = (items, columns, options = {}) => {
```
Since this function doesn't use any of the React's lifecycle let's not call it `useExportWithItems` it's just a helper function after all.

karelhala commented on packages/utils/src/useFilterConfig/filterChipHelpers.js:
```diff
@@ -0,0 +1,85 @@
+import { getFilterConfigItem } from './filterConfigHelpers';
+
+const chipsForGroup = (category, currentValue = {}, categoryConfig) =>
+  Object.keys(currentValue)
```
You could use `Object.entries` since we are accessing both key and value.

karelhala commented on packages/utils/src/useFilterConfig/filterChipHelpers.js:
```diff
@@ -0,0 +1,85 @@
+    .flatMap((groupValue) => {
+      const group = categoryConfig.items?.find(findWithString(groupValue));
+
+      return Object.keys(currentValue[groupValue]).map((itemValue) =>
```
Same as above `Object.entries` can be used here.

karelhala commented on packages/utils/src/useFilterConfig/filterChipHelpers.js:
```diff
@@ -0,0 +1,85 @@
+  ({
+    [conditionalFilterType.text]: (chip) => chip.chips[0].name[0],
+    [conditionalFilterType.checkbox]: (chip) => {
+      console.log(configItem, chip);
+      return 'YOLOO';
```
Hehe, I believe these are testing lines :smiley:

karelhala commented on packages/utils/src/useFilterConfig/useFilterConfig.js:
```diff
@@ -0,0 +1,37 @@
+
+  const onFilterDelete = async (_event, chips, clearAll = false) => {
+    clearAll ? (resetOnClear ? reset() : clear()) : deselect(...toDeselectValue(filterConfig, chips[0]));
+    onDeleteFilter && onDeleteFilter(chips, clearAll);
```
Safe accessor can be used here
```suggestion
    onDeleteFilter?.(chips, clearAll);
```

karelhala commented on packages/utils/src/useRowsBuilder/useRowsBuilder.js:
```diff
@@ -0,0 +1,81 @@
+  });
+
+// TODO separate out item handling into "WithItems" hook
+const useRowsBuilder = (items, columns, options = {}) => {
```
This doesn't use any of the React's lifecycle, let's omit the `use` word.

karelhala reviewed (commented):
Looking much better! I've got half way trough I don't want to spam you with change requests!

There's an empty file [packages/components/src/ColumnManager/columns.js](https://github.com/RedHatInsights/frontend-components/pull/1359/files#diff-faf0ed8a688d8b9926185f54c9ebfe1b19b5183a337a8a805d827f502ea6e382) is that correct?

bastilian commented:
@karelhala Thanks for taking another look! <3 There are still things I need to clean up and implement. I'll ping you once I think it is good!

bastilian commented on packages/utils/src/useTableTools/FilterConfigBuilder/ChipBuilder.js:
```diff
@@ -0,0 +1,96 @@
+
+const findWithString = (value) => (item) => String(item.value) === String(value);
+
+class ChipBuilder {
```
All these classes are now obsolete and turned into a set of helpers.

bastilian commented on packages/utils/src/useTableTools/FilterConfigBuilder/ChipBuilder.js:
```diff
@@ -0,0 +1,96 @@
+    const categoryConfig = this.filterConfig.getCategoryForLabel(filter);
+    const { label, type } = categoryConfig ? categoryConfig : { label: filter, type: null };
+
+    switch (type) {
```
There should be no more `switch`s in here and are replaces with `(x) = ({}[x])` kind of functions.

bastilian commented on packages/utils/src/useTableTools/helper.js:
```diff
@@ -0,0 +1,63 @@
+      return String(getSortable(property, a)).localeCompare(String(getSortable(property, b)));
+    } else {
+      return -String(getSortable(property, a)).localeCompare(String(getSortable(property, b)));
+    }
```
Incorporated.

bastilian commented on packages/utils/src/useTableTools/FilterConfigBuilder/FilterConfigBuilder.js:
```diff
@@ -0,0 +1,232 @@
+
+const defaultPlaceholder = (label) => `Filter by ${label.toLowerCase()}`;
+
+class FilterConfigBuilder {
```
Obsolete, the functionality is now implemented using helper functions

bastilian commented on packages/utils/src/useTableTools/FilterConfigBuilder/FilterConfigBuilder.js:
```diff
@@ -0,0 +1,232 @@
+  });
+
+  toFilterConfigItem = (item, handler, value) => {
+    switch (item.type) {
```
This pattern is now used all over instead of switch statements.

bastilian commented on packages/utils/src/useBulkSelect/helpers.js:
```diff
@@ -0,0 +1,40 @@
+  }
+};
+
+export const itemIds = (items) => items.map((item) => item.itemId);
```
Added.

bastilian commented on packages/utils/src/useBulkSelect/useBulkSelect.js:
```diff
@@ -0,0 +1,85 @@
+const useBulkSelect = ({ total, onSelect, preselected, itemIdsInTable, itemIdsOnPage, identifier = 'id' }) => {
+  const enableBulkSelect = !!onSelect;
+  // TODO use SelectionManager here.
+  const [selectedIds, setSelectedItemIds] = useState(preselected);
```
Added.

bastilian commented on packages/utils/src/useBulkSelect/useBulkSelect.js:
```diff
@@ -0,0 +1,85 @@
+  const onSelectCallback = async (func) => {
+    const newSelectedItemsIds = await func();
+    setSelectedItemIds(newSelectedItemsIds);
+    onSelect && onSelect(newSelectedItemsIds);
```
Added.

bastilian commented on packages/utils/src/useBulkSelect/useBulkSelect.js:
```diff
@@ -0,0 +1,85 @@
+
+  const unselectAll = () => [];
+  const selectNone = () => onSelectCallback(unselectAll);
+  const selectOne = (_, selected, key, row) => onSelectCallback(() => (selected ? selectItems([row[identifier]]) : unselectItems([row[identifier]])));
```
Added.

bastilian commented:
@karelhala This should finally be all good now and ready for another hopefully final review. I added some documentation and examples, I still have a some locally, but I'll submit that in another PR, since this one is already big. I tried to split the initial commits into adding each hook/function separately and following commits addressing the review comments.

karelhala commented on packages/utils/src/useTableTools/useTableTools.js:
```diff
@@ -0,0 +1,112 @@
+    setPage?.(1);
+  }, [JSON.stringify(activeFilters), JSON.stringify(sortBy)]);
+
+  return {
```
This is really heavy hook, I'd like to see some performance improvements to it later down the line.

karelhala commented on packages/utils/src/withExport/helpers.js:
```diff
@@ -0,0 +1,56 @@
+import camelCase from 'lodash/camelCase';
+import { downloadFile } from '../helpers';
+
+const CSV_FILE_PREFIX = 'compliance-export';
```
Can we change this to `CRC`?
```suggestion
const CSV_FILE_PREFIX = 'CRC';
```

karelhala reviewed (commented):
Looking really good! Just a couple of cosmetic changes and a couple of questions. But I think we're going in the right direction.

gkarat commented on packages/components/src/TableToolsTable/TableToolsTable.js:
```diff
@@ -0,0 +1,69 @@
+      transforms: propTypes.array,
+      sortByProperty: propTypes.string,
+      sortByArray: propTypes.array,
+      sortByFunction: propTypes.func,
```
And I would suggest changing this a bit (consistency nit-pick(:
```suggestion
      sortByFunc: propTypes.func,
```

gkarat commented on packages/components/src/TableToolsTable/TableToolsTable.js:
```diff
@@ -0,0 +1,69 @@
+import React from 'react';
```
Can we avoid the render of `input` container when there are no filters passed? By the `propTypes` spec, only `columns` and `items` required. 

![image](https://user-images.githubusercontent.com/31385370/180824675-a47727da-639b-42e9-8f62-b5f3e4db5ca4.png)

bastilian replied:
Nice catch! Fixed.

gkarat commented on packages/components/src/TableToolsTable/TableToolsTable.js:
```diff
@@ -0,0 +1,69 @@
+        <Pagination variant={PaginationVariant.bottom} {...toolbarProps.pagination} {...paginationProps} />
+      </TableToolbar>
+
+      {ColumnManager && <ColumnManager {...columnManagerProps} />}
```
I hear about this component for the first time. What is it used for?

bastilian replied:
This allows for, well, managing columns of tables. :)

https://user-images.githubusercontent.com/7757/182338632-21a983f0-151f-4fc0-9e39-e8499cc8d3dc.mp4

gkarat replied:
Cool stuff, thanks

gkarat commented on packages/utils/src/useTableTools/rowsBuilder.js:
```diff
@@ -0,0 +1,80 @@
+      {
+        title: () => <NoResultsTable />, // eslint-disable-line react/display-name
+        props: {
+          colSpan: 3,
```
Should this value be hardcoded? What if table has 4 columns?

bastilian replied:
The colSpan is now based on the columns.

gkarat commented on packages/utils/src/usePaginate/usePaginate.js:
```diff
@@ -0,0 +1,48 @@
+ *
+ * @param {Object} [options]
+ */
+const usePaginate = (options = {}) => {
```
Perhaps `usePagination` is a bit more clear naming

bastilian replied:
Renamed.

gkarat reviewed (approved):
Kudos, Sebastian, this is a great job 💯

I have left some comments on the things that caught my attention. This is impossible to review everything within this PR, however, I managed to render the component with some testing props, and played with the table doing basic operations like sorting, filtering, or selection.

I had some struggles identifying the right specification for each property. I really believe the gradual adoption of TS could really make the code navigation much easier and will serve as documentation for many unknown properties. But take all my suggestions as minor/for future work. Since this works as intended, and it will be easier to enhance the code with the next smaller PRs, I am approving this PR ✅ 

Another question coming to my mind: will this component work for both type of tables: those that use an API pagination/filtering/sorting (like Advisor for RHEL), and static table that fetches data only once (Advisor for OCP)?

bastilian commented on packages/components/src/ColumnManager/constants.js:
```diff
@@ -0,0 +1,5 @@
+import React from 'react';
```
Removed.

bastilian commented on packages/components/src/TableToolsTable/NoResultsTable.js:
```diff
@@ -0,0 +1,26 @@
+  <EmptyTable>
+    <Bullseye>
+      <EmptyState variant={EmptyStateVariant.full}>
+        <Title headingLevel="h5" size="lg">
```
Added.

bastilian commented on packages/components/src/TableToolsTable/TableToolsTable.js:
```diff
@@ -0,0 +1,52 @@
+      renderFunc: propTypes.func,
+    })
+  ).isRequired,
+  filters: propTypes.object,
+  options: propTypes.object,
+  toolbarProps: propTypes.object,
```
I'll refine the types when we move this to TS, which will be one of the next steps.

bastilian commented on packages/utils/src/useFilterConfig/helpers.js:
```diff
@@ -0,0 +1,37 @@
+
+export const itemForLabelInGroups = (configItem, label) => {
+  const flatItems = flattenConfigItems(configItem);
+  const item = flatItems.find(({ label: ItemLabel }) => {
+    return `${ItemLabel}` === `${label}`;
+  });
```
The reason, as far as I remember, this is done, because at some point there was an issue where a string with number for example for "8.1" would arrive here as an actually number and fail to the other, which would still be a string.

I'll note this as something to track down and iron out.

bastilian commented on packages/utils/src/useBulkSelect/useBulkSelect.js:
```diff
@@ -0,0 +1,90 @@
+        selectedIds,
+        selectNone: () => clear(),
+        tableProps: {
+          onSelect: total > 0 ? selectOne : undefined,
```
Yeah, we will need to allow for this at some point, probably, but for now, there is no need for it.

If we in the future do allow for passing down table or components props we should do so in a way that doesn't turn everything into a mess of spreads.

bastilian commented on packages/components/src/TableToolsTable/TableToolsTable.js:
```diff
@@ -0,0 +1,52 @@
+import useTableTools from '@redhat-cloud-services/frontend-components-utilities/useTableTools';
+
+const TableToolsTable = ({ items = [], columns = [], filters = [], options = {}, toolbarProps: toolbarPropsProp, ...tablePropsRest }) => {
+  const { toolbarProps, tableProps, ColumnManager } = useTableTools(items, columns, {
```
I'm not sure 100% what you mean... the `useTableTools` hook for now is just for the `TableToolsTable` component to not have all the hooks' code in here, but compacted in a hook. The only advantage of using it directly, would be to use the returns before passing them to (possible) custom table components. Which we will probably do when migrating to composed tables, as a stop gap.

bastilian commented:
@karelhala @Hyperkid123 Could we get your blessings here? Most of the review comments are addressed, but there are still things to improve, features to add, bugs to fix and documentation to add, as well as migrate the hooks to TypeScript. We can take better care of these things in smaller chunks once this is merged and the utilities are used by apps.

bastilian commented:
@karelhala @Hyperkid123 Can we please see that this can be merged? Even if (as discussed as a side discussion in the inventory refactor sync) we want to provide this via federated modules at some point, it still has to live somewhere. There are also already apps that have copied parts of these hooks and components, and it would be great if they could start using them from the FEC packages before they diverge (too much).

bastilian commented:
@karelhala @Hyperkid123 ping.

bastilian commented:
@karelhala @Hyperkid123 Ping. Again. Can we please either get a "No this will never be merged" or a "OK" to have it merged?

/cc @RedHatInsights/team-interact

johnsonm325 commented:
@karelhala @Hyperkid123 following up on behalf of @bastilian. It would be great to have this as Tasks app is already using this code locally in the Tasks app and it would be great to import it.

Are you both willing to merge this at some point?

jiridostal commented:
As this one has been open for a long time, I suggest merging it (as it does not modify anything that is existent, it should be safe) and applying additional changes in follow-up PRs. It would be very appreciated as there are apps waiting for this.

@ryelo there's been a couple of attempts to ping someone who can merge, can you please help us here?

ryelo commented:
@bastilian after talking to @jiridostal and the rest of Platform Experience. We're fine with these additions, but you'll need to create a new package in frontend-components and move those files there. We'll do that in order to allow your team to update and own the package.

karelhala reviewed (approved):
Looking good! Can you fix the conflicts?
